File: garnet/examples/fidl/echo_client_llcpp/echo_client.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fidl/examples/echo/llcpp/fidl.h>
#include <fuchsia/sys/cpp/fidl.h>
#include <lib/async-loop/cpp/loop.h>
#include <lib/async-loop/default.h>
#include <lib/async/default.h>
#include <lib/sys/cpp/component_context.h>
#include <lib/zx/channel.h>
#include <lib/zx/process.h>
#include <zircon/processargs.h>
#include <zircon/status.h>

#include <iostream>
#include <string>
#include <vector>

int main(int argc, const char** argv) {
  std::string server_url = "fuchsia-pkg://fuchsia.com/echo_server_llcpp#meta/echo_server_llcpp.cmx";
  std::string msg = "hello world";

  for (int i = 1; i < argc - 1; ++i) {
    if (!strcmp("--server", argv[i])) {
      server_url = argv[++i];
    } else if (!strcmp("-m", argv[i])) {
      msg = argv[++i];
    }
  }

  // Using high-level C++ bindings to launch the server component
  async::Loop loop(&kAsyncLoopConfigAttachToCurrentThread);
  ::fidl::InterfaceHandle<fuchsia::io::Directory> directory;
  fuchsia::sys::LaunchInfo launch_info;
  launch_info.url = server_url;
  launch_info.directory_request = directory.NewRequest().TakeChannel();
  fuchsia::sys::LauncherPtr launcher;
  auto context = sys::ComponentContext::Create();
  context->svc()->Connect(launcher.NewRequest());
  fuchsia::sys::ComponentControllerPtr controller;
  launcher->CreateComponent(std::move(launch_info), controller.NewRequest());

  sys::ServiceDirectory echo_provider(std::move(directory));
  zx::channel server_end, client_end;
  assert(zx::channel::create(0, &client_end, &server_end) == ZX_OK);
  echo_provider.Connect("fidl.examples.echo.Echo", std::move(server_end));

  // Using low-level C++ bindings to perform a call
  ::llcpp::fidl::examples::echo::Echo::SyncClient client(std::move(client_end));
  auto result = client.EchoString(fidl::StringView(msg));
  if (result.status() != ZX_OK) {
    std::cerr << "Failed to call server: " << result.status() << " (" << result.error() << ")"
              << std::endl;
    return result.status();
  }

  auto response = result.Unwrap();
  std::string reply_string(response->response.data(), response->response.size());
  std::cout << "Reply: " << reply_string << std::endl;

  return 0;
}

File: src/sys/component_manager/tests/events/components/chained_interpose/echo_client.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    anyhow::Error, fidl_fidl_examples_routing_echo as fecho,
    fidl_fuchsia_test_echofactory as fechofactory, fuchsia_async as fasync,
    fuchsia_component::client::connect_to_service,
};

#[fasync::run_singlethreaded]
async fn main() -> Result<(), Error> {
    let echo_factory = connect_to_service::<fechofactory::EchoFactoryMarker>()?;

    // Create the first echo channel.
    let (proxy_a, server_end_a) = fidl::endpoints::create_proxy::<fecho::EchoMarker>()?;
    echo_factory.request_echo_protocol(server_end_a).await?;
    proxy_a.echo_string(Some("a")).await?;

    // Create the second echo channel.
    let (proxy_b, server_end_b) = fidl::endpoints::create_proxy::<fecho::EchoMarker>()?;
    echo_factory.request_echo_protocol(server_end_b).await?;
    proxy_b.echo_string(Some("b")).await?;

    // Create the third echo channel.
    let (proxy_c, server_end_c) = fidl::endpoints::create_proxy::<fecho::EchoMarker>()?;
    echo_factory.request_echo_protocol(server_end_c).await?;
    proxy_c.echo_string(Some("c")).await?;

    Ok(())
}
